RZDataBinding/NSObject+RZDataBinding.m

```diff
@@ -412,28 +468,202 @@ - (instancetype)init
 {
     self = [super init];
     if ( self != nil ) {
-        self.observers = [NSPointerArray pointerArrayWithOptions:(NSPointerFunctionsWeakMemory | NSPointerFunctionsOpaquePersonality)];
+        _observers = [NSPointerArray pointerArrayWithOptions:(NSPointerFunctionsWeakMemory | NSPointerFunctionsOpaquePersonality)];
     }
     return self;
 }
 
 - (void)addObserver:(RZDBObserver *)observer
 {
-    [self.observers addPointer:(__bridge void *)(observer)];
+    @synchronized (self) {
+        [self.observers addPointer:(__bridge void *)(observer)];
+    }
 }
 
 - (void)removeObserver:(RZDBObserver *)observer
 {
-    __block NSUInteger observerIndex = NSNotFound;
-    [[self.observers allObjects] enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
-        if ( obj == observer ) {
-            observerIndex = idx;
-            *stop = YES;
+    @synchronized (self) {
+        __block NSUInteger observerIndex = NSNotFound;
+        [[self.observers allObjects] enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
+            if ( obj == observer ) {
+                observerIndex = idx;
+                *stop = YES;
+            }
+        }];
+
+        if ( observerIndex != NSNotFound ) {
+            [self.observers removePointerAtIndex:observerIndex];
+        }
+    }
+}
+
+@end
+
+#pragma mark - RZDBNotification implementation
+
+@implementation RZDBNotification
+
++ (instancetype)notificationForObserver:(RZDBObserver *)observer change:(NSMutableDictionary *)change
+{
+    RZDBNotification *notification = [[self alloc] init];
+    notification.target = observer.target;
+    notification.action = observer.action;
+    notification.queue = observer.callbackQueue;
+    notification.changeDict = change;
+
+    return notification;
+}
+
+- (BOOL)isEqual:(id)object
+{
+    @synchronized (self) {
+        BOOL equal = [super isEqual:object];
+
+        if ( !equal && [object isKindOfClass:[RZDBNotification class]] ) {
+            RZDBNotification *other = (RZDBNotification *)object;
+
+            // Do the callback queues need to be equal?
+            equal = (self.target == other.target) &&
+                    (self.action == other.action) &&
+                    (self.changeDict == other.changeDict ||
+                     (self.changeDict[kRZDBChangeKeyObject] == other.changeDict[kRZDBChangeKeyObject] &&
+                      [self.changeDict[kRZDBChangeKeyKeyPath] isEqualToString:other.changeDict[kRZDBChangeKeyKeyPath]]
+                     )
+                    );
+        }
+
+        return equal;
+    }
+}
+
+- (void)send
+{
+    @synchronized (self) {
+        if ( self.target != nil && self.action != NULL ) {
```

Comment:
`RZDBNotification` maintains a `weak` reference to the target, but `RZDBObserver` has the target `assign`. If the target goes away during the coalesce, the event shouldn't be fired.